### PR TITLE
[SG-33486] Periodically clean up old preview services

### PR DESCRIPTION
**Pull request by gitstart-sourcegraph**
## Description

- Create bash script to clean up inactive render.com service
- Create github action schedule event to execute the script

## Test plan

Make sure all checks are passed in CI

## Refs

- [SourceGraph Issue](https://github.com/sourcegraph/sourcegraph/issues/33486)
- [GitStart ticket](https://app.gitstart.com/clients/sourcegraph/tickets/SG-33486)

### CODE REVIEW

File: .github/workflows/pr-preview.yml [hunk 1]
```diff
@@ -4,8 +4,18 @@ on:
...
     types: [opened, synchronize, closed]
 
+  schedule:
+    # At minute 0 past every 12th hour
+    - cron: '0 */12 * * *'
+
+  # For testing
+  push:
+    branches:
+      - contractors/SG-33486
```
Comment (gitstart-sourcegraph) [inline .github/workflows/pr-preview.yml:14]: Will remove these before merging

File: .github/workflows/pr-preview.yml [hunk 2]
```diff
@@ -20,6 +30,22 @@ jobs:
         env:
           RENDER_COM_API_KEY: ${{ secrets.RENDER_COM_API_KEY }}
           RENDER_COM_OWNER_ID: ${{ secrets.RENDER_COM_OWNER_ID }}
-
           BUILDKITE_PULL_REQUEST_REPO: "https://github.com/${{ github.repository }}"
           BUILDKITE_BRANCH: ${{ github.head_ref }}
+
+  clean-up-inactive-pr-preview-apps:
+    if: github.event_name == 'schedule' || github.event_name == 'push'
```
Comment (gitstart-sourcegraph) [inline .github/workflows/pr-preview.yml:37]: will remove `github.event_name == 'push'` before merging

File: dev/ci/render-pr-preview-cleanup.sh [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  last_deploy_created_at_ISO="${last_deploy_created_at//.[[:digit:]]*Z/Z}"
+
+  if [[ $last_deploy_created_at_ISO < $expiration_date_ISO ]]; then
+    # there are no deploy times since `expiration_date_ISO` => remove it
+    echo "-- Removing service ${service_id} (last deployed at ${last_deploy_created_at})"
+
+    # curl -sSf -o /dev/null --request DELETE \
+    #   --url "https://api.render.com/v1/services/${service_id}" \
+    #   --header 'Accept: application/json' \
+    #   --header "Authorization: Bearer ${render_api_key}"
```
Comment (gitstart-sourcegraph) [inline dev/ci/render-pr-preview-cleanup.sh:88]: Uncomment this before merging/final test to make sure it works

**Comment by gitstart-sourcegraph** [comment]
@valerybugakov  could you review this spec PR? These are list of inactive render.com preview app which are listed in GitHub action: https://github.com/sourcegraph/sourcegraph/runs/6012307431?check_suite_focus=true

File: .github/workflows/pr-preview.yml [hunk 4]
```diff
@@ -20,6 +30,22 @@ jobs:
...
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 1
+
+      - name: Clean up inactive PR preview apps
+        # Clean up
+        run: ./dev/ci/render-pr-preview-cleanup.sh -e 5
```
Comment (valerybugakov) [inline .github/workflows/pr-preview.yml:48]: Could you comment here on the meaning of the script parameter `5`?

File: .github/workflows/pr-preview.yml [hunk 5]
```diff
@@ -20,6 +30,22 @@ jobs:
         env:
           RENDER_COM_API_KEY: ${{ secrets.RENDER_COM_API_KEY }}
           RENDER_COM_OWNER_ID: ${{ secrets.RENDER_COM_OWNER_ID }}
-
           BUILDKITE_PULL_REQUEST_REPO: "https://github.com/${{ github.repository }}"
           BUILDKITE_BRANCH: ${{ github.head_ref }}
+
+  clean-up-inactive-pr-preview-apps:
```
Comment (valerybugakov) [inline .github/workflows/pr-preview.yml:36]: Could you add a section about cleanup cron action to [the relevant documentation page](https://github.com/sourcegraph/sourcegraph/blob/main/doc/dev/how-to/client_pr_previews.md)?
To answer these questions:
1. Why is my preview inactive?
2. Where to find the script that cleans up previews?
3. What is the default cleanup schedule and the default preview lifetime?

File: dev/ci/render-pr-preview-cleanup.sh [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+# GET 100/minute
+# DELETE 30/minute
+
+cursor=""
+ids=()
+
+# Collect ids of all services which are created before expiration date and `not_suspended`
+for (( ; ; )); do
+  service_list=$(curl -sSf --request GET \
+    --url "https://api.render.com/v1/services?type=web_service&createdBefore=$(urlencode "$expiration_date_ISO")&suspended=not_suspended&ownerId=${render_owner_id}&limit=100&cursor=${cursor}" \
```
Comment (valerybugakov) [inline dev/ci/render-pr-preview-cleanup.sh:52]: After exploring render.com API it seems that services are updated on deployments. I checked 3 services using these two pages:

1. https://api-docs.render.com/reference/get-services
2. https://api-docs.render.com/reference/get-deploys

Results seem to confirm the assumption. What do you think about using `updatedBefore` to get relevant services right away without fetching deployments with a separate PR?

File: dev/ci/render-pr-preview-cleanup.sh [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+# We should take care about render.com rate limit (https://api-docs.render.com/reference/rate-limiting)
+# GET 100/minute
+# DELETE 30/minute
+
+cursor=""
+ids=()
+
+# Collect ids of all services which are created before expiration date and `not_suspended`
+for (( ; ; )); do
+  service_list=$(curl -sSf --request GET \
```
Comment (valerybugakov) [inline dev/ci/render-pr-preview-cleanup.sh:51]: Let's add links to relevant pages in the API reference documentation to ease further improvements of this script. E.g., https://api-docs.render.com/reference/get-services

This comment applies to all other render.com API requests.

File: dev/ci/render-pr-preview-cleanup.sh (hunk 6 above)
Comment (valerybugakov) [inline dev/ci/render-pr-preview-cleanup.sh:52]: It seems the services are updated on deployments. I check three services using these pages:

1. https://api-docs.render.com/reference/get-services
2. https://api-docs.render.com/reference/get-deploys

These examples confirm the assumption. What do you think about using the `updatedAt` filter to get relevant services without fetching `deloyments` separately?

Comment (gitstart-sourcegraph) [reply dev/ci/render-pr-preview-cleanup.sh]: Thanks @valerybugakov for this good catch. I updated the script with it

Comment (valerybugakov) [reply dev/ci/render-pr-preview-cleanup.sh]: @gitstart-sourcegraph, the update looks good to me.

File: doc/dev/how-to/client_pr_previews.md [hunk 8]
```diff
@@ -28,3 +28,9 @@ We do not create client PR previews [if Go or GraphQL is changed](https://source
 ## Why does a search query fail with an error?
 
 The preview app is deployed with `SOURCEGRAPHDOTCOM_MODE=false`, which means that the user should be authenticated to use all web application features similar to [the dogfood instance](https://k8s.sgdev.org/). Make sure that you're logged in. If it doesn't fix the issue, please report it in Slack.
+
+## Why is my preview inactive?
```
Comment (valerybugakov) [inline doc/dev/how-to/client_pr_previews.md:32]: @gitstart-sourcegraph, could you add answers to questions here?

File: .github/workflows/pr-preview.yml (hunk 5 above)
Comment (gitstart-sourcegraph) [reply .github/workflows/pr-preview.yml:31]: I updated the doc, could you please help to review and correct if it's needed

**Review by valerybugakov — APPROVED** [review]
LGTM! Let's wait for approval from the @sourcegraph/dev-experience team.

**Review by jhchabran — APPROVED** [review]
LGTM!

**Pull request merged**